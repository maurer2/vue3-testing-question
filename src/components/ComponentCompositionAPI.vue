<template>
  <section>
    <h1>Composition API</h1>
    <ButtonComponent
      id="button-1a"
      @click="handleClick1"
    />
    <ButtonComponent
      id="button-1b"
      @click="handleClick2()"
    />
  </section>
</template>

<script lang="ts">
import { ref, defineComponent } from 'vue';
import ButtonComponent from './ButtonComponent.vue';

export default defineComponent({
  name: 'ComponentCompositionAPI',
  components: {
    ButtonComponent,
  },
  setup() {
    const counter1 = ref<number>(0);
    const counter2 = ref<number>(0);

    function handleClick1(): void {
      counter1.value += 1;
    }

    function handleClick2(): void {
      counter2.value += 1;
    }

    return {
      counter1,
      counter2,
      handleClick1,
      handleClick2,
    };
  },
});

</script>

<style scoped>

</style>
