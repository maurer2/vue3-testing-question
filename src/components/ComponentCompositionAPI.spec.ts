// eslint-disable-next-line import/no-extraneous-dependencies
import { shallowMount } from '@vue/test-utils';
import ComponentCompositionAPI from './ComponentCompositionAPI.vue';

describe('Component Composition Api', () => {
  let cmp: any;

  beforeEach(() => {
    cmp = shallowMount(ComponentCompositionAPI, {});
  });

  afterEach(() => {
    cmp.unmount();

    jest.clearAllMocks();
    jest.resetModules();
  });

  it('renders', () => {
    expect(cmp.html()).toContain('Clicked');
  });

  it('button1 triggers counter update function when clicked v2', async () => {
    const spy = jest.spyOn(cmp.vm, 'handleClick1');

    await cmp.find('#button-1').trigger('click');
    expect(spy).toHaveBeenCalled();
  });

  it('button2 triggers counter update function when clicked', async () => {
    const spy = jest.spyOn(cmp.vm, 'handleClick2');

    await cmp.find('#button-2').trigger('click');
    expect(spy).toHaveBeenCalled();
  });

  it('button1 updates counter1', async () => {
    await cmp.find('#button-1').trigger('click');
    expect(cmp.vm.counter1).toBe(1);
  });

  it('button2 updates counter2', async () => {
    await cmp.find('#button-2').trigger('click');
    expect(cmp.vm.counter2).toBe(1);
  });
});
