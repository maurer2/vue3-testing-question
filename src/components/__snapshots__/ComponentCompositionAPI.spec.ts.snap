// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Composition Api matches snapshot 1`] = `
<div
  data-v-app=""
>
  <section>
    <h1>
      Composition API
    </h1>
    <button-component-stub
      id="button-1a"
    />
    <button-component-stub
      id="button-1b"
    />
  </section>
</div>
`;
